Correction of FrameUtility V3

moi15moi opened the pull request:
Finally, this should be the last version of FrameUtility.
For reference, here are the previous version: #37, #46

### **Why it is needed?**
The previous FrameUtility would not work with VFR video, now yes.
Finally, the previous FrameUtility was more a hack than anything else.

### **What has been done**
- Add TimeType Enum
- Add [ffmpeg](https://www.ffmpeg.org/) dependency to get the video file timestamps
- Add [mkvtoolnix](https://mkvtoolnix.download/) dependency to get the video file timestamps for mkv file since it is more performant then ffmpeg
- Be able to create Timestamps from 3 methods (``from_fps``, ``from_video_file``, ``from_timestamps_file``)
- Correction how to convert ms to ass_timestamps (``Convert.time``)
- Update of the algorithm of ``Convert.ms_to_frames`` and ``Convert.frames_to_ms``
- Add many tests to be sure that we have the right behaviour
- Update ``FrameUtility`` to use the ``Timestamps`` class.
- Add proof for the algorith of ms_to_frames since it isn't intuitive.

### **What still need to be been done**
- Correct the github action workflow. I don't know why, but when I try to run it with act, the workflow blocks and stop doing anything
- Update the logic of ``FrameUtility.add`` to use ``Timestamps``. I don't understand what is the expected behaviour of this method.
- Update pyonfx version
- Pack ffmpeg with pyonfx. We could try to replicate [this method](https://github.com/CoffeeStraw/VVVVID-Downloader/blob/master/.github/workflows/ci.yml#L36-L50)

CoffeeStraw commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,552 @@
+
+
+class Timestamps:
+    """Timestamps object contains informations about the timestamps of an video.
```
Moreover, users reading documentation need to know that they need to call one of the factory methods (from_fps, from_timestamps_file, from_video_file).

moi15moi replied:
Does this sound good to you?
```py
"""Timestamps object contains informations about the timestamps of an video.
Both Constant Frame Rate (CFR) and Variable Frame Rate (VFR) videos are supported.

To create an Timestamps object, you need to call one of the following method ``from_fps``, ``from_timestamps_file``, ``from_video_file``.
Ex:
	timestamps = Timestamps.from_fps(Fraction(24000, 1001))


Parameters:
	rounding_method (RoundingMethod): A rounding method. 99% of the time, you want to use RoundingMethod.ROUND.
		Note:
			Video player have 2 methods to deal with timestamps. Some floor them and other round them.
			This can lead to difference when displaying the subtitle.
				Ex:
					Player - Method - proof
					mpv    - round  - https://github.com/mpv-player/mpv/blob/7480efa62c0a2a1779b4fdaa804a6512aa488400/sub/sd_ass.c#L499
					FFmpeg - floor  - https://github.com/FFmpeg/FFmpeg/blob/fd1712b6fb8b7acc04ccaa7c02b9a5c9f233cfb3/libavfilter/vf_subtitles.c#L194-L196
					VLC    - floor  - https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/modules/codec/libass.c#L453-454
									  https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/include/vlc_tick.h#L120-132
					MPC-HC - floor  - https://github.com/clsid2/mpc-hc/blob/0994fd605a9fb4d15806d0efdd6399ba1bc5f984/src/Subtitles/LibassContext.cpp#L843
			Important note:
				Matroska (.mkv) file are an exception !!!
				If you want to be compatible with mkv, use RoundingMethod.ROUND.
				By default, they only have a precision to milliseconds instead of nanoseconds like most format.
					For more detail see:
						1- https://mkvtoolnix.download/doc/mkvmerge.html#mkvmerge.description.timestamp_scale
						2- https://matroska.org/technical/notes.html#timestampscale-rounding
	timestamps (List[int], optional): A list of [timestamps](https://en.wikipedia.org/wiki/Timestamp) in milliseconds encoded as integers.
							It represent each frame [presentation timestamp (PTS)](https://en.wikipedia.org/wiki/Presentation_timestamp)
	normalize (bool, optional): If True, it will shift the timestamps to make them start from 0. If false, the option does nothing.
	fpms (Fraction, optional): The fpms.
	last_frame_time (Fraction, optional): The last frame time not rounded.
"""
```

CoffeeStraw replied:
Better, but it's still not correct, because you created a more general method (and not only one that "contains information about the timestamps of a video").

moi15moi replied:
Could you update the doc for this?
I don't understand what you want

CoffeeStraw replied:
Sure, I will do it later

CoffeeStraw commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,552 @@
+        )
+
+    @staticmethod
+    def validate(timestamps: List[int]) -> None:
```
Since it doesn't do anything fancy, it seems to be called only in the `init` method and seems like the only place where it should be called, I would rather remove it and incorporate the 3 checks in the `init` to reduce the codebase.

moi15moi replied:
Done

CoffeeStraw commented on docs/Proof algorithm - ms_to_frames.md:
The math is correct, but we are missing the most important explanation: why, to obtain a frame, you need to add 0.5 to the ms BEFORE multiplying with fps * 1/1000. Why isn't it done just by:
```python
round(ms * fps * 1/1000)
```

moi15moi replied:
Which line are you talking about?
The general formula is: ``ms=method_of_rounding(frame * fps *1/1000)`` where ``method_of_rounding`` is ``floor`` or ``round``

CoffeeStraw replied:
Mhhhh... at a second glance, the formula isn't even right: you are talking about "ms_to_frames", but the formula is actually for "frames_to_ms"

That aside, my understanding is that the result of this formula yields a floating point number, but a `frame` is an integer. We must then decide which integer we should return.

In python3, there is the `round` method for that, but reading the code from the other (mpv/mpc-hc/libass...) it seems like everyone has opted for what you implemented as well for PyonFX. I can't understand why there was a need to implement a custom "round" and "floor" method.

moi15moi replied:
> Mhhhh... at a second glance, the formula isn't even right: you are talking about "ms_to_frames", but the formula is actually for "frames_to_ms"

Warning, I start with ``ms=method_of_rounding(frame * fps *1/1000)`` which is effectively ``frames_to_ms``, but from this equation, I isolate ``frame``, so it become ``ms_to_frames`` !

> In python3, there is the `round` method for that, but reading the code from the other (mpv/mpc-hc/libass...) it seems like everyone has opted for what you implemented as well for PyonFX. I can't understand why there was a need to implement a custom "round" and "floor" method.

[Python round](https://docs.python.org/3/library/functions.html#round) isn't the same has the mathematical one. Don't ask me why, I don't know.
Ex:
Python: round(0.5) = 0
Mathematic: round(0.5) = 1

This is why I implemented a custom round.
Now for the floor, simply to not import the one form math.floor. This is purely esthetic.

CoffeeStraw replied:
Mhhhh... Ok.
Then, what I would do is to expose our doubts in the python documentation.

We need to say that we just mimicked what other people did, and that we don't have the full knowledge about this matter. I'll do it later

moi15moi replied:
> Then, what I would do is to expose our doubts in the python documentation.

Which doubts are you talking about? The [round method](https://docs.python.org/3/library/functions.html#round) doesn't follow the mathematical rounding convention. See [this article](https://medium.com/thefloatingpoint/pythons-round-function-doesn-t-do-what-you-think-71765cfa86a8) for more information.

CoffeeStraw replied:
I was just wondering why the authors decided to go for a custom rounding method.
It doesn't seem to me that:
```python
# We use the upper bound
upper_bound = (ms + 0.5) * fps * 1/1000
# Then, we trunc the result
trunc_frame = int(upper_bound)

# If the upper_bound equals to the trunc_frame, this means that we don't respect the inequation because it is "greater than", not "greater than or equals".
# So if it happens, this means we need to return the previous frame
if upper_bound == trunc_frame:
    return trunc_frame - 1
else:
    return trunc_frame
```
does implement the mathematical rounding. To me it seems like just a... strange rounding.

The mathematical rounding for this case would just be:
```
int(ms * fps * 1/1000 + 0.5)
```

If you don't have an explanation as of why they implemented this rounding it's ok, but it seems important to me for it to be documented.

moi15moi replied:
> I was just wondering why the authors decided to go for a custom rounding method.

Because ``frames_to_ms`` needs to be exactly the opposite of ``ms_to_frames``. I am the authors of these equation (I mean, I didn't found them online. It is purely math deduction).
To calculate ``frames_to_ms``, it is well known that the equation is: $ms = roundingMethod(frame \times {1 \over fps} \times 1000)$.

Now, from the equation, our goal is to create a formula to be able to calculate an ``ms_to_frames`` function. To do so, we need to isolate $frame$ from the equation.

Since there are two possible roundingMethod options (floor, round), there are two different inequations that can be deduced from the original equation, which are written in the document.
Additionally, from the inequation, we can also deduce an algorithm




> The mathematical rounding for this case would just be:
> 
> ```
> int(ms * fps * 1/1000 + 0.5)
> ```

No, because $frame = round(ms \times fps \times {1 \over 1000})$ isn't perfectly the inverse of $ms = round(frame \times {1 \over fps} \times 1000)$. Important to note, here the rounding method round up, so, if it encounter $round(0.5)$, it will become $1$.

## Here is a proof by contradiction
$fps = 24000/1001$
$ms = 105 $
### With your method $frame = round(ms \times fps \times {1 \over 1000})$
```math
\begin{gather}
frame = round(ms \times fps \times {1 \over 1000}) \\
frame = round(105 \times {24000 \over 1001} \times {1 \over 1000}) \\
frame = round({360 \over 143}) \\
frame = 3
\end{gather}
```
### With what is wrotten in the proof doc
```math
\begin{gather}
upperBound = (ms + 0.5) \times fps * {1 \over 1000} \\
upperBound = {2532 \over 1001} \\
truncFrame = int(upperBound ) = 2 \\
 frame = truncFrame = 2
\end{gather}
```

### Little reminder to show which frame correspond to which ms
```math
\begin{gather}
fps = 24000/1001 \\
Frame_0 : [0, 42[ ms \\
Frame_1 : [42, 83[ ms \\
Frame_2 : [83, 125[ ms \\
Frame_3 : [125, 167[ ms
\end{gather}
```

### Conclusion
Like you can see, your method return $frame = 3$, but the one in the doc return $frame = 2$ which is the right result.

CoffeeStraw commented on pyonfx/convert.py:
```diff
@@ -63,6 +95,8 @@ def time(ass_ms: Union[int, str]) -> Union[str, int, ValueError]:
         # Milliseconds?
         if type(ass_ms) is int and ass_ms >= 0:
+            # From https://github.com/Aegisub/Aegisub/blob/6f546951b4f004da16ce19ba638bf3eedefb9f31/libaegisub/include/libaegisub/ass/time.h#L32
+            ass_ms = (ass_ms + 5) - (ass_ms + 5) % 10
```
I can't understand why we need to reduce precision, especially in such a delicate method.

moi15moi replied:
It rounds the milliseconds to centiseconds but the ASS format only supports centiseconds, so it doesn't reduce the precision.

Why should we do it?
It helps to have better precision.
Example: 49 ms is closest to 50 ms, but currently, it would floor it to 40 ms.

CoffeeStraw replied:
Got it, amazing job :)
I never noticed it.

We should just document this fact better, maybe by adding an extra line of comment

moi15moi replied:
Done

CoffeeStraw commented on pyonfx/convert.py:
```diff
@@ -82,76 +116,125 @@ def time(ass_ms: Union[int, str]) -> Union[str, int, ValueError]:
     @staticmethod
-    def ms_to_frames(ms: int, fps: Union[int, float, Fraction], is_start: bool) -> int:
-        """Converts from milliseconds to frames.
+    def ms_to_frames(
```
I would change the order of the parameters to:
- ms
- time_type
- timestamps

Then, should we keep the "approximate"? Why would anyone want to not "approximate" if it is beyond the video end (which should happen mostly for the last line only)?

moi15moi replied:
It helps the user to know, "Oops, anything above this point won't be rendered anyway, so why would I lose my time to process X line."

CoffeeStraw replied:
But then, if it is "True" by default I assume that most of the users will never touch this value, thus they will never get to know.

I am in favour of ease of use in cases like this, even if it to sacrifice a bit of utility: I would remove it and keep the approximation as the only behaviour. It seems to me that whoever will want, can discover and avoid to process (if needed) lines exceeding the end of the video.

moi15moi replied:
Done

CoffeeStraw commented on pyonfx/convert.py:
```diff
@@ -82,76 +116,125 @@ def time(ass_ms: Union[int, str]) -> Union[str, int, ValueError]:
+        return bisect.bisect_right(timestamps.timestamps, ms) - 1
 
     @staticmethod
     def frames_to_ms(
```
Follow comment above.

moi15moi replied:
See my comment above

moi15moi replied:
Done

CoffeeStraw commented on pyonfx/utils.py:
```diff
@@ -143,95 +142,109 @@ class FrameUtility:
-
-    This way, the frames' mid time will always be one of the player's seek times.
-    The only exception is the first frame, since we can't have negative times.
+    Both Constant Frame Rate (CFR) and Variable Frame Rate (VFR) videos are supported.
```
I would put this comment into the `Timestamps` documentation instead of here.

moi15moi replied:
Done

CoffeeStraw reviewed (changes requested):
I reviewed the general logic of the code, suggested some structural changes and exposed some doubts that I couldn't figure out myself.

Once this is solved I can take a look at the details of the implementation, at the documentation and at the tests.

moi15moi commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,552 @@
+        Ex:
+            Player - Method - proof
+            mpv    - round  - https://github.com/mpv-player/mpv/blob/7480efa62c0a2a1779b4fdaa804a6512aa488400/sub/sd_ass.c#L499
+            FFmpeg - floor  - https://github.com/FFmpeg/FFmpeg/blob/fd1712b6fb8b7acc04ccaa7c02b9a5c9f233cfb3/libavfilter/vf_subtitles.c#L195
```
I guess that you are talking about ffmpeg.
Like you can see [here](https://github.com/libass/libass/blob/9f4e6afeec54c0d240a9b8377f27e6c5b7c2bde0/libass/ass.h#L605-L606), the type of ``now`` is ``long long``, but ffmpeg call it with a ``double``. This means it will be truncated which is equivalent to floor.

CoffeeStraw replied:
Gotcha
Then it is probably more appropriate to put:
https://github.com/FFmpeg/FFmpeg/blob/fd1712b6fb8b7acc04ccaa7c02b9a5c9f233cfb3/libavfilter/vf_subtitles.c#L194-L196

moi15moi replied:
I updated the doc with your link

moi15moi commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,552 @@
+            Player - Method - proof
+            mpv    - round  - https://github.com/mpv-player/mpv/blob/7480efa62c0a2a1779b4fdaa804a6512aa488400/sub/sd_ass.c#L499
+            FFmpeg - floor  - https://github.com/FFmpeg/FFmpeg/blob/fd1712b6fb8b7acc04ccaa7c02b9a5c9f233cfb3/libavfilter/vf_subtitles.c#L195
+            VLC    - floor  - https://github.com/videolan/vlc/blob/f7bb59d9f51cc10b25ff86d34a3eff744e60c46e/include/vlc_tick.h#L118-L130
```
First of all, I should have used the gitlab repos, not github since I think their principal git repos is gitlab.
They call ass_render_frame [here](https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/modules/codec/libass.c#L453-454) which call [MS_FROM_VLC_TICK](https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/include/vlc_tick.h#L120-132)

I don't fully understand what a tick is, but here is my understanding: A [tick](https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/include/vlc_tick.h#L36-46) is a unit of time greater than milliseconds (seems to be microseconds). As you may know, double have a floating-point precision problem, so I assume VLC uses a higher unit to achieve better precision. Now, as for why it floors, it's because both variables (``vtk`` and ``CLOCK_FREQ``) are integers, which means the result will be truncated.

CoffeeStraw replied:
Gotcha, seems legit :)

moi15moi replied:
I updated VLC link in the documentation
Hope your ok with that

moi15moi commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,552 @@
+        timestamps (List[int], optional): A list of [timestamps](https://en.wikipedia.org/wiki/Timestamp) in milliseconds encoded as integers.
+                                It represent each frame [presentation timestamp (PTS)](https://en.wikipedia.org/wiki/Presentation_timestamp)
+        normalize (bool, optional): If True, it will shift the timestamps to make them start from 0. If false, the option does nothing.
+        fpms (Fraction, optional): The fpms.
```
I did that because ``from_timestamps_file`` and ``from_video_file`` deal with milliseconds.
But, we can change ``fpms`` to ``fps`` in the constructor, but it will create unnecessary calculations
Ex:
``from_timestamps_file`` will create a ``fpms``. It will convert it to ``fps`` then when it call the constructor, it will reconvert it to ``fpms``

CoffeeStraw replied:
Ahhhh, I think it is more because the methods "ms_to_frames" and "frames_to_ms" works with ms. So we can have everything in ms this way. We can keep it as you did, I would just specify what "fpms" is (frames per milliseconds) directly in the doc.

moi15moi replied:
Done

moi15moi commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,552 @@
+            99% of the time, you want to use RoundingMethod.ROUND.
+        timestamps (List[int], optional): A list of [timestamps](https://en.wikipedia.org/wiki/Timestamp) in milliseconds encoded as integers.
+                                It represent each frame [presentation timestamp (PTS)](https://en.wikipedia.org/wiki/Presentation_timestamp)
+        normalize (bool, optional): If True, it will shift the timestamps to make them start from 0. If false, the option does nothing.
```
At 99% of the time, the user want the normalize timestamps.
With you idea, this means the user would need to do 2 calls VS 1.
Ex:
Before
```py
timestamps = Timestamps.from_video_file(video_path) # by default, it normalize it
```

After
```py
timestamps = Timestamps.from_video_file(video_path)
timestamps.normalize()
```

Do you still want me to add your idea?

CoffeeStraw replied:
Mhhh... I think I am not getting the idea of normalizing then, mostly for VFR videos: if the frame-rate is different in different parts of the video, doesn't normalizing the timestamps yield a bad result? As we will misalign the timestamps with the video.

moi15moi replied:
> Mhhh... I think I am not getting the idea of normalizing then, mostly for VFR videos:

But I am not sure if you totally understand what the method does.
It only shift the timestamps so the first one start at 0.

> doesn't normalizing the timestamps yield a bad result?

Yes and no. This is a very very tricky subject that depend on the video player (i am pretty sure that they don't act all the same, but I don't have any proof of that)

Honestly, I don't have the level to explain properly why it is necessary. I just know that sometimes, the video player will normalized them depending on the timestamps of the audio and/or chapter.

Just for you to know, [Aegisub](https://github.com/arch1t3cht/Aegisub/blob/752a14cc25f361b2c5265d93561fe997e78d576d/libaegisub/common/vfr.cpp#L48-L53) always normalize the video timestamps.

99% of the video are already normalize.

CoffeeStraw replied:
Mhhhh... Do you know what MPV does in this respect?

moi15moi replied:
From what I know, MPV will take the first timestamps of the video or audio or chapter and normalize from that.

**Ex 1 - All timestamps equals:**
video first timestamps = 83 ms
audio first timestamps = 83 ms
chapter first timestamps = 83 ms
Then, it will set the video first timestamps to 0.

**Ex 2 - Different timestamps:**
video first timestamps = 83 ms
audio first timestamps = 83 ms
chapter first timestamps = 0 ms
Then, it will set the video first timestamps to 83

CoffeeStraw replied:
Ok. Your solution seems legit then.
Thank you for the explanation.

moi15moi commented on .github/workflows/ci.yml:
```diff
@@ -65,11 +65,15 @@ jobs:
+      - name: Install platform-specific requirements (Windows)
+        if: matrix.os == 'windows-latest'
+        run: |
+          choco install ffmpeg -y
```
I removed the code with ``Chocolatey`` in the workflow, but I want to clarify that it was only to run the workflow, not to build the package.

CoffeeStraw commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,552 @@
+from typing import Callable, List, Optional, Tuple, Union
+
+
+class RoundingMethod(Enum):
+    FLOOR: Callable[[Fraction], int] = lambda ms: int(ms)
+    ROUND: Callable[[Fraction], int] = lambda ms: int(ms + Fraction("0.5"))
```
There is no need to define them as callable, as we never call them. Simply keeping them as enum should be enough.

moi15moi replied:
We call them.

See:
- [convert.py - line 206](https://github.com/CoffeeStraw/PyonFX/blob/242f7a4d87bc283f807c03b898a49616203d0ed6/pyonfx/convert.py#L206)
- [timestamps.py - line 184](https://github.com/CoffeeStraw/PyonFX/blob/242f7a4d87bc283f807c03b898a49616203d0ed6/pyonfx/timestamps.py#L184)
- [timestamps.py - line 189](https://github.com/CoffeeStraw/PyonFX/blob/242f7a4d87bc283f807c03b898a49616203d0ed6/pyonfx/timestamps.py#L189)
- [timestamps.py - line 193](https://github.com/CoffeeStraw/PyonFX/blob/242f7a4d87bc283f807c03b898a49616203d0ed6/pyonfx/timestamps.py#L193)
- [timestamps.py - line 242](https://github.com/CoffeeStraw/PyonFX/blob/242f7a4d87bc283f807c03b898a49616203d0ed6/pyonfx/timestamps.py#L242)
- [timestamps.py - line 474](https://github.com/CoffeeStraw/PyonFX/blob/242f7a4d87bc283f807c03b898a49616203d0ed6/pyonfx/timestamps.py#L474)

JeremieBergeron commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,555 @@
+        """
+
+        if os.path.isfile(path_to_timestamps_file_or_content):
+            with open(path_to_timestamps_file_or_content, "r") as f:
```
We may need to specify the encoding to "utf-8" because timestamps can contains comments with non-ascii character.

[mkvtoolnix](https://gitlab.com/mbunkus/mkvtoolnix/-/blob/c6911337e4019702eff27758f89ee47a88b8f269/src/common/mm_io.cpp#L51-71) doesn't seems to care about the encoding. It will just convert the bytes to char, so equivalent to ascii.

But if we set the encoding to "ascii", it will just raise an exception when the file contains a non-ascii char, so it would be safer to use utf-8

What do you think about it?

moi15moi replied:
We may need to specify the encoding to "utf-8" because timestamps can contains comments with non-ascii character.

[mkvtoolnix](https://gitlab.com/mbunkus/mkvtoolnix/-/blob/c6911337e4019702eff27758f89ee47a88b8f269/src/common/mm_io.cpp#L51-71) doesn't seems to care about the encoding. It will just convert the bytes to char, so equivalent to ascii.

But if we set the encoding to "ascii", it will just raise an exception when the file contains a non-ascii char, so it would be safer to use utf-8

What do you think about it?

CoffeeStraw commented:
Had another quick glance at the code: do you think we could avoid uploading a new font (thus making use of the font we already fetch in the workflow)?

Also, how much does it take to generate the fake videos? Could we generate them on the fly while testing to keep the codebase minimal?

moi15moi commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,681 @@
+
+        if ffprobe_output_dict["format"]["format_name"] == "matroska,webm":
+            # We only do this check for .mkv file. See the note about mkv in the class documentation
+            time_base = Fraction(ffprobe_output_dict["streams"][0]["time_base"])
```
Currently, the ``_from_mkvextract`` method doesn't perform a check for the time_base like ffmpeg because when I wrote this code, it was a information that mkvtoolnix didn't expose.
But, I asked if it could be added [here](https://help.mkvtoolnix.download/t/how-to-get-timestamp-scale-information-without-using-mkvinfo).
It was introduced in version 82.0, which was released on 2 January 2024. We could perform a version check and if the user has a V82 or higher, then we perform the check.

Also, I propose to add a new file named ``mkvtoolnix.py`` and extract the timestamps in that file since ``timestamps.py`` is becoming too big.

moi15moi commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,681 @@
+            video_path,
+            "-J",
+        ]
+        mkvmerge_output = subprocess.run(cmd, capture_output=True, text=True)
```
Add ``--command-line-charset UTF-8 --output-charset UTF-8`` to the command. It is necessary when the user asks for a filepath that contains special characters that the locale encoding (which be queried with ``locale.getencoding()``) doesn't support.
```suggestion
        cmd = [
                mkvmerge_path,
                video_path,
                "-J",
                "--command-line-charset", "UTF-8",
                "--output-charset", "UTF-8",
        ]
        mkvmerge_output = subprocess.run(cmd, capture_output=True, text=True, encoding="utf-8")
```

moi15moi commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,681 @@
+            "0:" + timestamps_file_path
+        ]
+
+        mkvextract_output = subprocess.run(cmd, capture_output=True, text=True)
```
Same problem has [here](https://github.com/CoffeeStraw/PyonFX/pull/48/files#r1503115505)

moi15moi commented on pyonfx/timestamps.py:
```diff
@@ -0,0 +1,681 @@
+            "-print_format",
+            "json",
+        ]
+        ffprobe_output = subprocess.run(cmd, capture_output=True, text=True)
```
ffprobe seems to always output the result in utf-8 has said [here](https://trac.ffmpeg.org/ticket/8890#comment:5).
```suggestion
        ffprobe_output = subprocess.run(cmd, capture_output=True, text=True, encoding="utf-8")
```

moi15moi commented:
I push some changes.

- I removed the [ms_to_frames](https://pyonfx.readthedocs.io/en/latest/reference/convert.html#pyonfx.convert.Convert.ms_to_frames), [frames_to_ms](https://pyonfx.readthedocs.io/en/latest/reference/convert.html#pyonfx.convert.Convert.frames_to_ms) and [move_ms_to_frame](https://pyonfx.readthedocs.io/en/latest/reference/convert.html#pyonfx.convert.Convert.move_ms_to_frame) method.
Now, the user should use [VideoTimestamps](https://github.com/moi15moi/VideoTimestamps/) because the conversion weren't always perfect before.
- Now, the constructor of [FrameUtility](https://pyonfx.readthedocs.io/en/latest/reference/utils.html#pyonfx.utils.FrameUtility) use [``ABCTimestamps``](https://github.com/moi15moi/VideoTimestamps/blob/25de23c3c025156c3dfbaae903844402bafd9310/video_timestamps/abc_timestamps.py) instead of a ``fps``. The user will be able to pass a [FPSTimestamps](https://github.com/moi15moi/VideoTimestamps/blob/25de23c3c025156c3dfbaae903844402bafd9310/video_timestamps/fps_timestamps.py), [VideoTimestamps](https://github.com/moi15moi/VideoTimestamps/blob/25de23c3c025156c3dfbaae903844402bafd9310/video_timestamps/video_timestamps.py) or a [TextFileTimestamps](https://github.com/moi15moi/VideoTimestamps/blob/25de23c3c025156c3dfbaae903844402bafd9310/video_timestamps/text_file_timestamps.py) which should cover any use case.
- I updated all the packages used in the CI.

moi15moi commented on pyonfx/utils.py:
```diff
@@ -143,68 +143,79 @@ class FrameUtility:
     Example:
-        >>> FU = FrameUtility(0, 110, 20)
+        >>> # Let's assume to have an Ass object named "io" having a 20 fps video (i.e. frames are 50 ms long)
+        >>> FU = FrameUtility(0, 110, io.input_timestamps)
```
If I remember correctly, you wanted to have a ``io.input_timestamps`` object.
Is it still needed?

CoffeeStraw replied:
Hi Moi! Thank you so much for the work, I've jus seen the changes and they're looking really good!
Yes, I think it would be better to have it computed automatically and saved in the io object (we can also allow disabling it through a flag for optimization if wanted).

Once this is done, the PR is looking pretty good already, I'll just go through a read and we're ready!

moi15moi replied:
Done, let me know what you think.

CoffeeStraw commented on pyonfx/ass_core.py:
```diff
@@ -486,6 +504,17 @@ def get_media_abs_path(mediafile):
+                            Path(self.meta.video), video_index
+                        )
+                    else:
+                        fps_str = line[11:].strip().split(":")[1]
```
Why line 11? Doesn't appear to be safe for all the .ass files

moi15moi replied:
For the same reason as [this](https://github.com/moi15moi/PyonFX/blob/8e0b61a3479c49e6ab89bfc601c9ceea93dc5511/pyonfx/ass_core.py#L504), to skip ``Video File:``.
Also, this part is covered by the test [``test_ass_values``](https://github.com/CoffeeStraw/PyonFX/blob/8e0b61a3479c49e6ab89bfc601c9ceea93dc5511/tests/test_ass.py#L142-L145).

CoffeeStraw replied:
Ops :)
The original code is so old I didn't remember.

Makes sense then! Let me have a final read and I'll get this through!

CoffeeStraw reviewed (commented):
Looking good! I have just a doubt on the computed timestamps value. Left a comment for it!

moi15moi commented on pyonfx/utils.py:
```diff
@@ -143,68 +143,79 @@ class FrameUtility:
+        - This ensures the subtitle will be visible for the entire frame duration
+
+        Frame timings example:
+        Frame # : Start - End (Mid-point)
```
We talked in pm.
The commit https://github.com/CoffeeStraw/PyonFX/pull/48/commits/6dc3a2f4bcd4df7612c3761cddc9bb3d20f6d5c6 fix this issue.
